Wabbajack.Networking.Http/ResumableDownloader.cs

```diff
@@ -1,187 +1,146 @@
-        return await file.Hash(token);
     }
 
-    private DownloadConfiguration CreateConfiguration(HttpRequestMessage message)
+    private async Task<AbsolutePath> DownloadStreamDirectlyToFile(HttpRequestMessage message, AbsolutePath filePath, IJob job, CancellationToken token)
     {
-        var maximumMemoryPerDownloadThreadMb = Math.Max(0, _performanceSettings.MaximumMemoryPerDownloadThreadMb);
-        var configuration = new DownloadConfiguration
-        {
-            MaximumMemoryBufferBytes = maximumMemoryPerDownloadThreadMb * 1024 * 1024,
-            Timeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds,
-            ReserveStorageSpaceBeforeStartingDownload = true,
-            RequestConfiguration = new RequestConfiguration
-            {
-                Headers = message.Headers.ToWebHeaderCollection(),
-                ProtocolVersion = message.Version,
-                UserAgent =  message.Headers.UserAgent.ToString()
-            }
-        };
+        if(job.Size == null) throw new ArgumentException("Job size must be set before downloading");
 
-        return configuration;
-    }
+        using Stream fileStream = GetDownloadFileStream(filePath);
 
-    private void OnDownloadFileCompleted(object? sender, AsyncCompletedEventArgs e)
-    {
-        _error = e.Error;
-    }
+        var startingPosition = fileStream.Length;
 
-    private async void OnDownloadProgressChanged(object? sender, DownloadProgressChangedEventArgs e)
-    {
-        var processedSize = e.ProgressedByteSize;
-        if (_job.Current == 0)
-        {
-            // Set current to total in case this download resumes from a previous one
-            processedSize = e.ReceivedBytesSize;
-        }
+        _logger.LogDebug("Download for '{name}' is starting from {position}...", filePath.FileName.ToString(), startingPosition);
+
+        var httpClient = _httpClientFactory.CreateClient("ResumableClient");
+
+        message.Headers.Range = new RangeHeaderValue(startingPosition, null);
+        using var response = await httpClient.SendAsync(message, HttpCompletionOption.ResponseHeadersRead, token);
 
-        await _job.Report(processedSize, _token);
-        if (_job.Current > _job.Size)
+        var responseContentLength = response.Content.Headers.ContentLength;
+
+        if (responseContentLength != null && responseContentLength > fileStream.Length)
         {
-            // Increase job size so progress doesn't appear stalled
-            _job.Size = (long)Math.Floor(_job.Current * 1.1);
+            fileStream.SetLength(responseContentLength.Value);
         }
-    }
 
-    private void OnDownloadStarted(object? sender, DownloadStartedEventArgs e)
-    {
-        _job.ResetProgress();
+        var responseStream = await response.Content.ReadAsStreamAsync(token);
 
-        if (_job.Size < e.TotalBytesToReceive)
+        long reportProgressThreshold = 100 * 1024 * 1024; // Report progress every 100MB
```

Comment:
Smart, those small files don't really matter anyway. However 100MB might be kind of large still, for those with slower connections. Maybe better to lower this to say 10MB?

Reply:
Adjusted.

I think there's some good improvements that could be made to managing settings, could easily make that an adjustable value if we valued it enough.